app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt

```diff
@@ -1,12 +1,269 @@
+
+        fun makeScope(): MutableMap<String, NamedNode> {
+            return HashMap<String, NamedNode>().toMutableMap()
+        }
+
+        fun addName(name: String, node: NamedNode, scope: MutableMap<String, NamedNode>) {
+            scope[name] = node
+            if (!visibleNames.containsKey(name))
+                visibleNames[name] = NameOverloadState()
+            if (node is Variable || node is Function.Parameter)
+                visibleNames[name]!!.addVariable(node)
+            if (node is Function)
+                visibleNames[name]!!.addFunction(node)
+        }
+
+        fun destroyScope(scope: MutableMap<String, NamedNode>) {
+            for ((name, node) in scope) {
+                // if (name, node) exists in scope then the node has to be on top of the corresponding NameOverloadState stack
+                if (node is Variable || node is Function.Parameter)
+                    visibleNames[name]!!.popVariable()
+                if (node is Function)
+                    visibleNames[name]!!.popFunction()
+
+                // invariant: no node of a particular name exists <==> visibleNames[name] does not exist
+                if (visibleNames[name]!!.isEmpty())
+                    visibleNames.remove(name)
+            }
+        }
+
+        // Core function
+
+        fun analyzeNode(
+            node: Any,
+            currentScope: MutableMap<String, NamedNode>
+        ) {
+
+            when (node) {
+
+                is Program -> {
+                    val newScope = makeScope()
+
+                    // we have to be able to map "napisz(...)" FunctionCalls to something
+                    // TODO: napiszNode has to actually have some meaning
+                    val dummyNapiszNode: Function = Function(
+                        "napisz",
+                        listOf(Function.Parameter("wartość", Type.Number, null)),
+                        Type.Unit,
+                        listOf()
+                    )
+                    addName("napisz", dummyNapiszNode, newScope)
+
+                    node.globals.forEach { analyzeNode(it, newScope) }
+                    destroyScope(newScope)
+                }
+
+                is Program.Global.VariableDefinition -> {
+                    reportIfNameConflict(node.variable.name, currentScope)
+
+                    // first analyze the variable, then add name, because we can't have self-referencing definitions
+                    analyzeNode(node.variable, currentScope)
+                    addName(node.variable.name, node.variable, currentScope)
+                }
+
+                is Program.Global.FunctionDefinition -> {
+                    reportIfNameConflict(node.function.name, currentScope)
+
+                    // first add name, then analyze because we can have recursive calls in the body
+                    addName(node.function.name, node.function, currentScope)
+                    analyzeNode(node.function, currentScope)
```

Comment:
There is a small problem here: we allow self-reference in default parameter values: `czynność f(x: Czy = f())`

Reply:
Right, I think I can move adding names from `FunctionDefinition` to `Function` and from `VariableDefinition` to `Variable`